### PR TITLE
[Frontend] Delay visibility of progress bar

### LINKED ISSUE #1168: [Notifications] Progress dialogs may flash
Originally mentioned in [this comment](https://github.com/nasa/openmct/issues/588#issuecomment-244422876), an indeterminate progress dialog may "flash" briefly when saving. This is most observable/noticeable when using Save and Continue Editing, but can also be seen with regular Save (esp. with a remote persistence store.)

This is undesirable; if a Save can be completed very quickly, the dialog is not useful and instead serves only as a distraction or annoyance.

### PR DESCRIPTION
**Pull request by hudsonfoo**
Resolves #1168

### CODE REVIEW

**Comment by charlesh88** [comment]
@hudsonfoo Hi Hudson! A few notes on this one for you:
- We definitely don't want to apply this globally to the .overlay class - that would mean all dialogs would be delayed, such as Save As..., Move, etc. The should be done as a dedicated class, something like `.delayEntry100ms`. We'll apply this class to the 'in progress' instance of the overlay (forthcoming notes on this from @akhenry) and we can also reuse it with anything else we want to delay for that time amount. I think the best spot for that class would be in /platform/commonUI/general/res/sass/_effects.scss.
- We should do this via the animation-delay property instead of using an animation with duration. I'd use two keyframes (0 and 100%), keep the anim dur very short (50ms perhaps or less) and use animation-delay of 100ms.
- Both those values (the duration and the delay value itself) to be added and used as new constants, something like `$durEntryMs` and `$delayEntryMs`. The values should be added to /platform/commonUI/general/res/sass/_constants.scss.
- Your animation shouldn't need to (and shouldn't) declare `display: block` on each line. My understanding is that the element is injected into the DOM when it's ready to display, so it's not like we have to change from `display: none` to make it appear. Was there a reason you needed to include that?
- Why is `z-index: 5` included? We shouldn't need to mess with the layering for this solution, yes?

As always, thanks for the great contributions, and please ping us with any questions or issues.

**Comment by akhenry** [comment]
@hudsonfoo I am convinced that CSS is the right approach now, thanks. 

We do need to make this a little more specific, and I think the way to do this it to add an option to dialogs to allow them to optionally be delayed. I think that the best way of doing this is by adding a 'delay' property to the [DialogModel](https://github.com/nasa/openmct/blob/master/platform/commonUI/dialog/src/DialogService.js#L194) type that is passed into [DialogService.showBlockingMessage](https://github.com/nasa/openmct/blob/master/platform/commonUI/dialog/src/DialogService.js#L229)

It would have to be a boolean that, if provided, passes the delay class mentioned by @charlesh88  through to the overlay service in [this line](https://github.com/nasa/openmct/blob/master/platform/commonUI/dialog/src/DialogService.js#L89).

Let me know if you have any questions.

**Comment by hudsonfoo** [comment]
@charlesh88 Agreed on all points. Stayed up extra late coding and made a couple of obvious mistakes I now regret :)

> - Why is z-index: 5 included? We shouldn't need to mess with the layering for this solution, yes?

I really tried to stay away from modifying z-index but I was having the strangest issue without it. Basically if the opacity of the overlay was less than 100%, the `.object-holder` elements would show above the `.overlay` and the `.overlay` showed above everything else. The moment it hit 100% everything looked correct. Z-index of 5 fixes that issue and didn't appear to cause any regressions but I'm still not certain what was going on there. Will avoid using z-index in the next update and try to drill down into the cause.

@akhenry I can do that, no prob.

**Comment by hudsonfoo** [comment]
🎱 It is decidedly so.

**Comment by hudsonfoo** [comment]
Btw, the build is failing, but not for any of the code I've changed. Is this a regression on master or am I missing something?

**Comment by VWoeltjen** [comment]
Hmm, [last build on master](https://circleci.com/gh/nasa/openmct/2689) succeeded, and there's nothing in this PR that should cause the build to fail as it did. I see other branches have failures as well. I think the version bump between es6-promise 3.2.2 and 3.3.0 changed something; filed #1183 and will investigate further.

**Comment by VWoeltjen** [comment]
#1190 should have fixed the build failure, so if you merge from master now I would not expect to see Circle CI failures. Thanks!

**Comment by hudsonfoo** [comment]
All set. Thanks for looking into that.

File: platform/commonUI/dialog/res/templates/overlay.html [hunk 1]
```diff
@@ -19,7 +19,7 @@
  this source code distribution or the Licensing information page available
  at runtime from the About dialog for additional information.
 -->
-<div class="abs overlay">
+<div class="abs overlay" ng-class="{'delayEntry100ms' : !!ngModel.delay}">
```
Comment (akhenry) [inline platform/commonUI/dialog/res/templates/overlay.html]: I'm a simple man, and double negative logic hurts my brain. Angular will handle the `undefined` case gracefully, so I'd be inclined to remove the `!!`

File: platform/commonUI/general/res/sass/overlay/_overlay.scss [hunk 2]
```diff
@@ -19,8 +19,17 @@
...
+
 .overlay {
     font-size: 90%;
+    &.delayEntry100ms {
+        @include keyframes(fadeInFromNone) {
+            0% { display: none; opacity: 0; }
+            100% { display: block; opacity: 1; }
+        }
+        @include animation-delay($delayEntryMs);
+        @include animation(fadeInFromNone $durEntryMs ease-in);
```
Comment (akhenry) [inline platform/commonUI/general/res/sass/overlay/_overlay.scss]: CSS is not my forte, do we need a duration of 100ms here as well as the delay of 100ms?

**Comment by akhenry** [comment]
One question, and one suggestion inline, other than that this looks good.

@charlesh88 can your peepers at the Sass changes, and give a go/no-go to merge?

**Comment by charlesh88** [comment]
Bella! Emoticon for two thumbs up! Go for it.

File: platform/commonUI/general/res/sass/overlay/_overlay.scss (hunk 2 above)
Comment (hudsonfoo) [inline platform/commonUI/general/res/sass/overlay/_overlay.scss]: @akhenry I just pushed an update to shorten the duration to 50ms which was requested a few comments back. Are you wanting that to be shorter or zero? I'm good either way.

File: platform/commonUI/dialog/res/templates/overlay.html (hunk 1 above)
Comment (hudsonfoo) [inline platform/commonUI/dialog/res/templates/overlay.html]: Understood :) Removed the double negative.

**Comment by akhenry** [comment]
@hudsonfoo Sorry for the delay in merging this, we've been in code freeze the last couple of days for the end of a release, but we're out now. If you can stick an author checklist on this, I will go ahead and merge it.

**Comment by hudsonfoo** [comment]
- [X] Changes address original issue?
- [X] Unit tests included and/or updated with changes?
- [X] Command line build passes?
- [X] Changes have been smoke-tested?

**Comment by hudsonfoo** [comment]
Thanks @akhenry!

**Comment by akhenry** [comment]
### Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? N/A stylesheet changes only
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y

**Pull request merged**